In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
# for dirname, _, filenames in os.walk('/kaggle/input'):
#     for filename in filenames:
#         print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/home/irfan/.local/share/virtualenvs/ds_code-Sb-xMZMk/lib/python3.9/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


# Ken Jee's YouTube Channel

# Summary code 
In this notebook I very lightly explore the data. The goal of this is to get you familiar with the basic features in the data so that you can start exploring it on your own. 

In [2]:
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
#read in data 
df_vid = pd.read_csv('Aggregated_Metrics_By_Video.csv')
df_vid_subs = pd.read_csv('Aggregated_Metrics_By_Country_And_Subscriber_Status.csv.zip')
df_ts = pd.read_csv('Video_Performance_Over_Time.csv.zip')
df_comments = pd.read_csv('All_Comments_Final.csv.zip')

In [4]:
#look at columns for each dataframe 
print(df_vid.columns)
print(df_vid_subs.columns)
print(df_ts.columns)
print(df_comments.columns)


Index(['Video', 'Video title', 'Video pub­lish time', 'Com­ments ad­ded',
       'Shares', 'Dis­likes', 'Likes', 'Sub­scribers lost',
       'Sub­scribers gained', 'RPM (USD)', 'CPM (USD)',
       'Av­er­age per­cent­age viewed (%)', 'Av­er­age view dur­a­tion',
       'Views', 'Watch time (hours)', 'Sub­scribers',
       'Your es­tim­ated rev­en­ue (USD)', 'Im­pres­sions',
       'Im­pres­sions click-through rate (%)'],
      dtype='object')
Index(['Video Title', 'External Video ID', 'Video Length', 'Thumbnail link',
       'Country Code', 'Is Subscribed', 'Views', 'Video Likes Added',
       'Video Dislikes Added', 'Video Likes Removed',
       'User Subscriptions Added', 'User Subscriptions Removed',
       'Average View Percentage', 'Average Watch Time', 'User Comments Added'],
      dtype='object')
Index(['Date', 'Video Title', 'External Video ID', 'Video Length',
       'Thumbnail link', 'Views', 'Video Likes Added', 'Video Dislikes Added',
       'Video Likes Removed', 'User Sub

### 1st row contains some aggregates, that may be confusing for now

In [5]:
df_vid = df_vid[1:]
df_vid.head(2)

Video                                        Video title  \
1  4OZip0cgOho  How I Would Learn Data Science (If I Had to St...   
2  78LjdAAw0wA                  100K Channel Update + AMA Stream!   

  Video pub­lish time  Com­ments ad­ded  Shares  Dis­likes  Likes  \
1         May 8, 2020               907    9583        942  46903   
2        Nov 12, 2020               412       4          4    130   

   Sub­scribers lost  Sub­scribers gained  RPM (USD)  CPM (USD)  \
1                451                46904      6.353     12.835   
2                 15                   12      2.668      6.259   

   Av­er­age per­cent­age viewed (%) Av­er­age view dur­a­tion    Views  \
1                              36.65                   0:03:09  1253559   
2                               6.26                   0:05:14     2291   

   Watch time (hours)  Sub­scribers  Your es­tim­ated rev­en­ue (USD)  \
1          65850.7042         46453                          7959.533   
2            200.2966            -3                             6.113   

   Im­pres­sions  Im­pres­sions click-through rate (%)  
1       26498799                                  3.14  
2         188318                                  0.72

In [6]:
df_vid.dtypes

Video                                    object
Video title                              object
Video pub­lish time                      object
Com­ments ad­ded                          int64
Shares                                    int64
Dis­likes                                 int64
Likes                                     int64
Sub­scribers lost                         int64
Sub­scribers gained                       int64
RPM (USD)                               float64
CPM (USD)                               float64
Av­er­age per­cent­age viewed (%)       float64
Av­er­age view dur­a­tion                object
Views                                     int64
Watch time (hours)                      float64
Sub­scribers                              int64
Your es­tim­ated rev­en­ue (USD)        float64
Im­pres­sions                             int64
Im­pres­sions click-through rate (%)    float64
dtype: object

### Let's rename columns to some short and suitable names

In [7]:
df_vid = df_vid.rename(columns={'Video':'id', 'Video title':'title', 'Video pub­lish time':'date', 
                        'Com­ments ad­ded':'comments','Shares':'shares', 'Dis­likes':'dislikes',
                        'Likes':'likes','Sub­scribers lost':'lost_subs',
                        'Sub­scribers gained':'gained_subs','RPM (USD)':'rpm', 'CPM (USD)':'cpm',
                        'Av­er­age per­cent­age viewed (%)':'average_view_pct',
                        'Av­er­age view dur­a­tion':'average_view_time','Views':'views',
                        'Watch time (hours)':'watch_time','Sub­scribers':'subscribers',
                        'Your es­tim­ated rev­en­ue (USD)':'estimated_revenue','Im­pres­sions':'impressions',
                        'Im­pres­sions click-through rate (%)':'CTRate'})

In [8]:
videos = df_vid
videos.columns

Index(['id', 'title', 'date', 'comments', 'shares', 'dislikes', 'likes',
       'lost_subs', 'gained_subs', 'rpm', 'cpm', 'average_view_pct',
       'average_view_time', 'views', 'watch_time', 'subscribers',
       'estimated_revenue', 'impressions', 'CTRate'],
      dtype='object')

# Problem Set / Questions to be Answered

- What topics get the most viewership?
- What do are people asking for in the comments?
- What impacts watch time and click through rate?
- Can we predict if a title will be clickable? 
- What thumbnails are most appealing (thumbnail data available in df_agg)
- What is different about my "viral" videos and normal videos
- Who is the core audience of my channel?

### What 'Aggregated_Metrics_By_Videos.csv' data set tells?

In [9]:
videos.head(2)

id                                              title  \
1  4OZip0cgOho  How I Would Learn Data Science (If I Had to St...   
2  78LjdAAw0wA                  100K Channel Update + AMA Stream!   

           date  comments  shares  dislikes  likes  lost_subs  gained_subs  \
1   May 8, 2020       907    9583       942  46903        451        46904   
2  Nov 12, 2020       412       4         4    130         15           12   

     rpm     cpm  average_view_pct average_view_time    views  watch_time  \
1  6.353  12.835             36.65           0:03:09  1253559  65850.7042   
2  2.668   6.259              6.26           0:05:14     2291    200.2966   

   subscribers  estimated_revenue  impressions  CTRate  
1        46453           7959.533     26498799    3.14  
2           -3              6.113       188318    0.72

the data set tells for every video:
- comments it gained
- how much it was shared
- how much it was liked and disliked
- subscribers
- view stats
- click-through rate

How many videos?

In [10]:
videos.shape

(223, 19)

## Q1- What topics get the most viewership?

# Making Observations

the data set contains some vids with -ve subscribers, lets filter them out

In [11]:
sub_videos = videos[videos['subscribers']>0]
sub_videos.shape

(216, 19)

In [12]:
(216/223)*100

96.8609865470852

96.9% videos did gain subscribers.

In [13]:
sub_videos = sub_videos.sort_values(by='views')
sub_videos.describe()

comments       shares    dislikes         likes   lost_subs  \
count  216.000000   216.000000  216.000000    216.000000  216.000000   
mean    61.601852   183.240741   17.967593   1038.115741   13.333333   
std     87.874322   744.525732   70.651216   3631.886949   33.720879   
min      2.000000     0.000000    0.000000     18.000000    0.000000   
25%     19.000000    22.000000    1.000000    168.000000    2.000000   
50%     37.000000    48.000000    4.000000    358.500000    7.000000   
75%     65.250000   115.500000   11.000000    722.500000   14.000000   
max    907.000000  9583.000000  942.000000  46903.000000  451.000000   

        gained_subs         rpm         cpm  average_view_pct         views  \
count    216.000000  216.000000  215.000000        216.000000  2.160000e+02   
mean     606.129630    4.503509   11.533986         34.350046  2.569858e+04   
std     3353.954989    1.763560    3.321150         14.375554  9.096247e+04   
min        2.000000    0.000000    5.439000          7.320000  4.550000e+02   
25%       29.000000    3.320750    9.481000         23.975000  4.044500e+03   
50%       76.500000    4.348500   11.302000         35.500000  8.671500e+03   
75%      260.500000    5.427500   12.974500         43.462500  1.852425e+04   
max    46904.000000   10.387000   37.786000         75.620000  1.253559e+06   

         watch_time   subscribers  estimated_revenue   impressions      CTRate  
count    216.000000    216.000000         216.000000  2.160000e+02  216.000000  
mean    1466.255796    592.796296         134.410125  4.639384e+05    3.095324  
std     4876.113005   3321.804153         562.989055  1.893355e+06    1.567778  
min        6.308900      1.000000           0.000000  1.863500e+04    0.490000  
25%      187.472575     22.750000          13.063250  9.947100e+04    1.965000  
50%      420.520800     66.000000          34.363000  1.542310e+05    2.905000  
75%     1157.035775    243.500000          97.246500  3.005855e+05    3.992500  
max    65850.704200  46453.000000        7959.533000  2.649880e+07    8.620000

75th percentile is 43.5 in avg view %age, 1.852425e+04 in views, 1157.035775 in watch_time

In [14]:
sub_videos[sub_videos['views']>18524].shape

(54, 19)

out of 223 videos 59 lie in 75th percentile, and 53 had more than 18.5k views

In [15]:
sub_videos[sub_videos['views']>18524]['likes'].max()

46903

In [16]:
from collections import defaultdict

In [17]:
words = defaultdict(int)
 
# memoizing count
for sub in sub_videos[sub_videos['views']>18000]['title']:
    for wrd in sub.split():
        words[wrd] += 1
 
# getting max frequency
for i in words:
    if words[i] >10:
        print(i, words[i])

to 11
a 14
Data 58
Science 43
- 11
How 17
I 13


Let's see what the release dates of most viewed videos tell?

In [18]:
sub_videos['date'] = pd.to_datetime(sub_videos['date'])

In [19]:
sub_videos[sub_videos['views']>18000]['date'].min()

Timestamp('2018-11-14 00:00:00')

In [20]:
sub_videos[sub_videos['views']>18000]['date'].max()

Timestamp('2021-12-27 00:00:00')

In [21]:
DS_vids = sub_videos[sub_videos['title'].str.contains('Data Science')]
DS_vids['title'].count()

143

From November 2018 to December 2021 there are 143 videos on Data Science. About 64% videos are on Data Science, in total data set.

In [22]:
DS_vids[DS_vids['views']>18000]['title'].count()

44

out of 143 Data Science videos, 44 had more than 18000 views.

Let's see what is the date range for these videos

In [23]:
DS_vids[DS_vids['views']>18000]['date'].min()

Timestamp('2018-11-14 00:00:00')

In [24]:
DS_vids_copy = DS_vids.copy()
DS_vids_copy['date'] = DS_vids_copy['date'].dt.to_period('Q')
DS_vids_copy = DS_vids_copy.groupby('date')

In [25]:
DS_vids_copy.count()

id  title  comments  shares  dislikes  likes  lost_subs  gained_subs  \
date                                                                           
2018Q4   2      2         2       2         2      2          2            2   
2019Q1   7      7         7       7         7      7          7            7   
2019Q2   6      6         6       6         6      6          6            6   
2019Q3   9      9         9       9         9      9          9            9   
2019Q4   9      9         9       9         9      9          9            9   
2020Q1  21     21        21      21        21     21         21           21   
2020Q2  27     27        27      27        27     27         27           27   
2020Q3  29     29        29      29        29     29         29           29   
2020Q4  10     10        10      10        10     10         10           10   
2021Q1   6      6         6       6         6      6          6            6   
2021Q2   8      8         8       8         8      8          8            8   
2021Q3   3      3         3       3         3      3          3            3   
2021Q4   5      5         5       5         5      5          5            5   
2022Q1   1      1         1       1         1      1          1            1   

        rpm  cpm  average_view_pct  average_view_time  views  watch_time  \
date                                                                       
2018Q4    2    2                 2                  2      2           2   
2019Q1    7    7                 7                  7      7           7   
2019Q2    6    6                 6                  6      6           6   
2019Q3    9    9                 9                  9      9           9   
2019Q4    9    9                 9                  9      9           9   
2020Q1   21   21                21                 21     21          21   
2020Q2   27   26                27                 27     27          27   
2020Q3   29   29                29                 29     29          29   
2020Q4   10   10                10                 10     10          10   
2021Q1    6    6                 6                  6      6           6   
2021Q2    8    8                 8                  8      8           8   
2021Q3    3    3                 3                  3      3           3   
2021Q4    5    5                 5                  5      5           5   
2022Q1    1    1                 1                  1      1           1   

        subscribers  estimated_revenue  impressions  CTRate  
date                                                         
2018Q4            2                  2            2       2  
2019Q1            7                  7            7       7  
2019Q2            6                  6            6       6  
2019Q3            9                  9            9       9  
2019Q4            9                  9            9       9  
2020Q1           21                 21           21      21  
2020Q2           27                 27           27      27  
2020Q3           29                 29           29      29  
2020Q4           10                 10           10      10  
2021Q1            6                  6            6       6  
2021Q2            8                  8            8       8  
2021Q3            3                  3            3       3  
2021Q4            5                  5            5       5  
2022Q1            1                  1            1       1

most Data Science videos were released in year 2020

In [26]:
DS_vids[(DS_vids['date']>('2019-12-31'))  & (DS_vids['date']<('2021-01-01')) & (DS_vids['views']==482)]

id                                            title       date  \
213  foggudmDDvY  Data Science in Golf: PGA Merchandise Show 2020 2020-01-31   

     comments  shares  dislikes  likes  lost_subs  gained_subs    rpm     cpm  \
213         6       5         0     18          0            2  3.529  12.105   

     average_view_pct average_view_time  views  watch_time  subscribers  \
213             38.65           0:03:31    482     28.3566            2   

     estimated_revenue  impressions  CTRate  
213              1.701        41477    0.69

In [27]:
DS_vids[(DS_vids['date']>('2019-12-31'))  & (DS_vids['date']<('2021-01-01')) & (DS_vids['views']==1253559)]

id                                              title       date  \
1  4OZip0cgOho  How I Would Learn Data Science (If I Had to St... 2020-05-08   

   comments  shares  dislikes  likes  lost_subs  gained_subs    rpm     cpm  \
1       907    9583       942  46903        451        46904  6.353  12.835   

   average_view_pct average_view_time    views  watch_time  subscribers  \
1             36.65           0:03:09  1253559  65850.7042        46453   

   estimated_revenue  impressions  CTRate  
1           7959.533     26498799    3.14

What about videos without Data Science?

In [28]:
NDS_vids = sub_videos[~sub_videos['title'].str.contains('Data Science')]
NDS_vids['title'].count()

73

In [29]:
NDS_vids[NDS_vids['views']>=18000]['title'].count()

13

In [31]:
DS_videos = 143
DS_videos_18k = 44
non_DS_videos = 73
non_DS_videos_18k = 13

print("18K+ videos' percentage of Data Science Videos: ", round((DS_videos_18k/DS_videos)*100))
print("18K+ videos' percentage of Non-Data Science Videos: ", round((non_DS_videos_18k/non_DS_videos)*100))

18K+ videos' percentage of Data Science Videos:  31
18K+ videos' percentage of Non-Data Science Videos:  18


# Some Observations about Videos Data Set

- **there are total 223 videos**
- **216/223 videos got subscribers**
- **54/223 videos lie in the 75th percentile and made 18k+ views**
- **143/223 (60%) videos are related to *Data Science***
- **44/143 *Data Science* videos got 18k+ views**
- **73/223 videos did not had Data Science in their title**
- **13/73 Non-Data Science videos got 18k+ views**
- **in these 54 videos *Data Science* was most commonly used in title**